packages/sns/lib/utils/snsSubscriber.ts

```diff
@@ -19,14 +24,23 @@ export async function subscribeToTopic(
   sqsClient: SQSClient,
   snsClient: SNSClient,
   queueConfiguration: CreateQueueCommandInput,
-  topicConfiguration: CreateTopicCommandInput,
+  topicConfiguration: TopicResolutionOptions,
   subscriptionConfiguration: SNSSubscriptionOptions,
   extraParams?: ExtraSNSCreationParams & ExtraSQSCreationParams & ExtraParams,
 ) {
-  const topicArn = await assertTopic(snsClient, topicConfiguration, {
-    queueUrlsWithSubscribePermissionsPrefix: extraParams?.queueUrlsWithSubscribePermissionsPrefix,
-    allowedSourceOwner: extraParams?.allowedSourceOwner,
-  })
+  let topicArn = isSNSTopicLocatorType(topicConfiguration) ? topicConfiguration.topicArn : undefined
+
+  if (!topicArn) {
+    if (isCreateTopicCommand(topicConfiguration)) {
+      topicArn = await assertTopic(snsClient, topicConfiguration, {
+        queueUrlsWithSubscribePermissionsPrefix:
+          extraParams?.queueUrlsWithSubscribePermissionsPrefix,
+        allowedSourceOwner: extraParams?.allowedSourceOwner,
+      })
+    } else {
+      topicArn = await getTopicArnByName(snsClient, topicConfiguration.topicName)
```

Comment:
If it's not either create command or locate command, shouldn't we throw an error here?

Reply:
yeah, this code is complex and hard to follow. would be nice to break into into more reasonable pieces

Reply:
I read it again more carefully. Actually it's part of the "locate" logic, but instead of having exact topic arn , we use `topicName` to resolve it into `topicArn`. 

Yes, I believe at least "extract method" could be applied here. I'm on it.

Reply:
Done, code adjusted.